pkg/mutation/system_test.go

```diff
@@ -301,25 +301,193 @@ func TestMutation(t *testing.T) {
+		}
+		if gotErr != nil {
+			t.Fatalf("got Mutate() error = %v, want <nil>", gotErr)
+		}
+	})
+}
+
+func TestSystem_DontApplyConflictingMutationsRemoveOriginal(t *testing.T) {
+	// Two conflicting mutators.
+	foo := &fakeMutator{
+		MID:    types.ID{Name: "foo"},
+		MPath:  mustParse("spec.containers[name: foo].image"),
+		GVKs:   []schema.GroupVersionKind{{Version: "v1", Kind: "Pod"}},
+		Labels: map[string]string{"active": "true"},
+	}
+	fooConflict := &fakeMutator{
+		MID:    types.ID{Name: "foo-conflict"},
+		MPath:  mustParse("spec.containers.image"),
+		GVKs:   []schema.GroupVersionKind{{Version: "v1", Kind: "Pod"}},
+		Labels: map[string]string{"active": "true"},
+	}
+
+	// Put System in an inconsistent state.
+	s := NewSystem()
+	err := s.Upsert(foo)
+	if err != nil {
+		t.Fatalf("got Upsert() error = %v, want <nil>", err)
+	}
+	err = s.Upsert(fooConflict)
+	if err == nil {
+		t.Fatalf("got Upsert() error = %v, want error", err)
+	}
+	gotErr := s.Remove(types.ID{Name: "foo"})
+	if gotErr != nil {
+		t.Fatalf("got Remove() error = %v, want <nil>", gotErr)
+	}
+
+	u := &unstructured.Unstructured{}
+	gotMutated, gotErr := s.Mutate(u, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "billing"}})
+	// TODO(willbeason): Fix once System is updated to properly report conflicts (#1216).
+	if gotMutated {
+		t.Errorf("got Mutate() = %t, want true", gotMutated)
+	}
+	// if !gotMutated {
+	// 	t.Errorf("got Mutate() = %t, want true", gotMutated)
+	// }
+	if gotErr != nil {
+		t.Fatalf("got Mutate() error = %v, want <nil>", gotErr)
+	}
+}
+
+func TestSystem_ConflictingMutationsDontImpactNonconflicting(t *testing.T) {
+	// Two conflicting mutators.
+	foo := &fakeMutator{
+		MID:    types.ID{Name: "foo"},
+		MPath:  mustParse("spec.containers[name: foo].image"),
+		GVKs:   []schema.GroupVersionKind{{Version: "v1", Kind: "Pod"}},
+		Labels: map[string]string{"active": "true"},
+	}
+	fooConflict := &fakeMutator{
+		MID:    types.ID{Name: "foo-conflict"},
+		MPath:  mustParse("spec.containers.image"),
+		GVKs:   []schema.GroupVersionKind{{Version: "v1", Kind: "Pod"}},
+		Labels: map[string]string{"active": "true"},
+	}
+	// A non-conflicting mutator on the same type.
+	bar := &fakeMutator{
+		MID:    types.ID{Name: "foo"},
+		MPath:  mustParse("spec.images[name: nginx].tag"),
+		GVKs:   []schema.GroupVersionKind{{Version: "v1", Kind: "Pod"}},
+		Labels: map[string]string{"active": "true"},
+	}
+
+	// Put System in an inconsistent state.
+	s := NewSystem()
+	err := s.Upsert(foo)
+	if err != nil {
+		t.Fatalf("got Upsert() error = %v, want <nil>", err)
+	}
+	err = s.Upsert(fooConflict)
+	if err == nil {
+		t.Fatalf("got Upsert() error = %v, want error", err)
+	}
+	err = s.Upsert(bar)
+	if err != nil {
+		t.Fatalf("got Upsert() error = %v, want <nil>", err)
+	}
+
+	u := &unstructured.Unstructured{}
+	gotMutated, gotErr := s.Mutate(u, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "billing"}})
```

Comment:
Should we also be checking that we're getting the mutation we are expecting here? That foo/fooConflict haven't somehow become un-conflicted due to the third mutator?

Reply:
Done.